### ISSUE TITLE
edgeHub Failed - DotNetty.Transport.Channels.ConnectTimeoutException

### ISSUE THREAD

**Issue by carloshm** [Azure/iotedge#343]
Several gateways are connected to tha same IoTHub. At some point some of those gateways start throwing errors for the connection, although even under lost of connection EdgeHub should handle it through offline behavior.

09/22/2018 12:31:28: [Error 501] Error in the module initialization.: DotNetty.Transport.Channels.ConnectTimeoutException: connection timed out: 172.18.0.2:8883
   at Microsoft.Azure.Devices.Client.InternalClient.<>c.<ApplyTimeout>b__62_2(Task t)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__276_1(Object obj)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at VibrationModule.Program.<Init>d__2.MoveNext() in /app/Program.cs:line 60

## Expected Behavior

Connection should not fail under timeout, and module should not be set inactive or offline should be redefined.

## Current Behavior

Connection is lost and after 40 hours it was not recovered until several edgeHub restart. edgeHub was not giving information about the process (setting modeul inactive), and giving erros after the restarts.

2018-09-21 11:57:49.068 +00:00 [WRN] - Closing connection for device: gateway/VibrationModule, Microsoft.Azure.Devices.ProtocolGateway.ProtocolGatewayException: Keep Alive timed out., 
2018-09-21 11:57:49.068 +00:00 [INF] - Disposing MessagingServiceClient for device Id xbio001pi01/VibrationModule because of exception - Microsoft.Azure.Devices.ProtocolGateway.ProtocolGatewayException: Keep Alive timed out.
2018-09-21 11:57:49.068 +00:00 [INF] - Setting device proxy inactive for device Id gateway/VibrationModule

## Steps to Reproduce

Running a module for long periods of time, with offline and reconnection slots.

### Device (Host) Operating System

Raspberry Pi Stretch

### Architecture

arm32

### Container Operating System

Linux

#### iotedged

iotedge 1.0.1 (6e5e86dcf0c9a3732fc72a64d9ec9b0fcb2d6fad)

#### Edge Agent

mcr.microsoft.com/azureiotedge-agent:1.0 

#### Edge Hub

mcr.microsoft.com/azureiotedge-hub:1.0

#### Docker

Client:
 Version:      18.06.0-dev
 API version:  1.37
 Go version:   go1.10.2
 Git commit:   daf021fe
 Built:        Wed Jun 27 00:53:52 2018
 OS/Arch:      linux/arm
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      dev
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.10.2
  Git commit:   9aac187
  Built:        Tue Jun 26 23:27:27 2018
  OS/Arch:      linux/arm
  Experimental: false

## Additional Information

SDK 1.18 https://github.com/Azure/azure-iot-sdk-csharp/releases/tag/2018-8-17 

**### EdgeHub logs**

[2018-09-22 12:41:31 +00:00]: Starting Edge Hub
[09/22/2018 12:41:33.030 PM] Edge Hub Main()
2018-09-22 12:41:53.394 +00:00 [INF] - Starting Edge Hub
2018-09-22 12:41:53.471 +00:00 [INF] - Version - 1.0.1.15962126 (6e5e86dcf0c9a3732fc72a64d9ec9b0fcb2d6fad)
2018-09-22 12:41:53.473 +00:00 [INF] - 
        █████╗ ███████╗██╗   ██╗██████╗ ███████╗
       ██╔══██╗╚══███╔╝██║   ██║██╔══██╗██╔════╝
       ███████║  ███╔╝ ██║   ██║██████╔╝█████╗
       ██╔══██║ ███╔╝  ██║   ██║██╔══██╗██╔══╝
       ██║  ██║███████╗╚██████╔╝██║  ██║███████╗
       ╚═╝  ╚═╝╚══════╝ ╚═════╝ ╚═╝  ╚═╝╚══════╝

 ██╗ ██████╗ ████████╗    ███████╗██████╗  ██████╗ ███████╗
 ██║██╔═══██╗╚══██╔══╝    ██╔════╝██╔══██╗██╔════╝ ██╔════╝
 ██║██║   ██║   ██║       █████╗  ██║  ██║██║  ███╗█████╗
 ██║██║   ██║   ██║       ██╔══╝  ██║  ██║██║   ██║██╔══╝
 ██║╚██████╔╝   ██║       ███████╗██████╔╝╚██████╔╝███████╗
 ╚═╝ ╚═════╝    ╚═╝       ╚══════╝╚═════╝  ╚═════╝ ╚══════╝

2018-09-22 12:41:53.474 +00:00 [INF] - Installing intermediate certificates.
2018-09-22 12:41:54.194 +00:00 [INF] - Attempting to connect to IoT Hub for client gateway/$edgeHub via AMQP...
2018-09-22 12:41:54.714 +00:00 [INF] - Connected to IoT Hub for client gateway/$edgeHub via AMQP, with client operation timeout 60000.
2018-09-22 12:41:54.771 +00:00 [INF] - Created cloud connection for client gateway/$edgeHub
2018-09-22 12:41:54.784 +00:00 [INF] - New cloud connection created for device gateway/$edgeHub
2018-09-22 12:41:54.836 +00:00 [INF] - Initializing configuration
2018-09-22 12:41:58.315 +00:00 [INF] - Created persistent store at /tmp/edgeHub
2018-09-22 12:41:58.526 +00:00 [INF] - Created new message store
2018-09-22 12:41:58.555 +00:00 [INF] - Started task to cleanup processed and stale messages
2018-09-22 12:41:59.078 +00:00 [INF] - New device connection for device gateway/$edgeHub
2018-09-22 12:42:06.285 +00:00 [INF] - Exiting disconnected state
2018-09-22 12:42:06.409 +00:00 [INF] - Entering connected state
2018-09-22 12:42:09.019 +00:00 [INF] - Obtained edge hub config from module twin
2018-09-22 12:42:10.047 +00:00 [INF] - Set the following 2 route(s) in edge hub
2018-09-22 12:42:10.055 +00:00 [INF] - managementToIoTHub: FROM /messages/modules/VibrationModule/outputs/ManagementToCloudOutput INTO $upstream
2018-09-22 12:42:10.055 +00:00 [INF] - uploadToIoTHub: FROM /messages/modules/UploadModule/outputs/ManagementToCloudOutput INTO $upstream
2018-09-22 12:42:10.062 +00:00 [INF] - Updated message store TTL to 14400 seconds
2018-09-22 12:42:10.065 +00:00 [INF] - Updated the edge hub store and forward configuration
2018-09-22 12:42:10.068 +00:00 [INF] - Initialized edge hub configuration
2018-09-22 12:42:10.593 +00:00 [INF] - Starting protocol heads - (MQTT, AMQP, HTTP)
2018-09-22 12:42:10.672 +00:00 [INF] - Starting MQTT head
2018-09-22 12:42:11.673 +00:00 [INF] - Initializing TLS endpoint on port 8883 for MQTT head.
2018-09-22 12:42:12.143 +00:00 [INF] - Starting AMQP head
2018-09-22 12:42:12.228 +00:00 [INF] - Started MQTT head
2018-09-22 12:42:12.979 +00:00 [INF] - Started AMQP head
2018-09-22 12:42:13.073 +00:00 [INF] - Starting HTTP head
2018-09-22 12:42:14.260 +00:00 [INF] - User profile is available. Using '"/home/edgehubuser/.aspnet/DataProtection-Keys"' as key repository; keys will not be encrypted at rest.
2018-09-22 12:42:17.199 +00:00 [WRN] - Overriding address(es) '"http://+:80"'. Binding to endpoints defined in "UseKestrel()" instead.
2018-09-22 12:42:17.457 +00:00 [INF] - Started HTTP head
2018-09-22 12:42:17.578 +00:00 [INF] - Attempting to connect to IoT Hub for client gateway/VibrationModule via AMQP...
2018-09-22 12:42:17.618 +00:00 [INF] - Connected to IoT Hub for client gateway/VibrationModule via AMQP, with client operation timeout 60000.
2018-09-22 12:42:17.622 +00:00 [INF] - Created cloud connection for client gateway/VibrationModule
2018-09-22 12:42:17.623 +00:00 [INF] - New cloud connection created for device gateway/VibrationModule
2018-09-22 12:42:17.829 +00:00 [INF] - New token requested by client gateway/VibrationModule, but using existing token as it is usable.
2018-09-22 12:42:17.839 +00:00 [INF] - Successfully generated identity for clientId gateway/VibrationModule and username gateway/gateway/VibrationModule/?api-version=2018-06-30&DeviceClientType=Microsoft.Azure.Devices.Client%2F1.18.0%20%28.NET%20Core%204.6.0.0%3B%20Linux%204.14.52-v7%2B%20%231123%20SMP%20Wed%20Jun%2027%2017%3A35%3A49%20BST%202018%3B%20Arm%29
2018-09-22 12:42:17.874 +00:00 [INF] - ClientAuthenticated, gateway/VibrationModule, 795a4b4a
2018-09-22 12:42:18.303 +00:00 [INF] - New device connection for device gateway/VibrationModule
2018-09-22 12:42:18.385 +00:00 [INF] - Bind device proxy for device gateway/VibrationModule
2018-09-22 12:42:18.391 +00:00 [INF] - Binding message channel for device Id gateway/VibrationModule
2018-09-22 12:42:19.509 +00:00 [INF] - Set subscriptions from session state for gateway/VibrationModule
2018-09-22 12:42:21.865 +00:00 [INF] - Set subscriptions from session state for gateway/VibrationModule
2018-09-22 12:42:21.991 +00:00 [INF] - Set subscriptions from session state for gateway/VibrationModule
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable

---------------------------------------------------------------------------------------------------------

[2018-09-22 12:51:01 +00:00]: Starting Edge Hub
[09/22/2018 12:51:04.549 PM] Edge Hub Main()
2018-09-22 12:51:34.299 +00:00 [INF] - Starting Edge Hub
2018-09-22 12:51:34.426 +00:00 [INF] - Version - 1.0.1.15962126 (6e5e86dcf0c9a3732fc72a64d9ec9b0fcb2d6fad)
2018-09-22 12:51:34.429 +00:00 [INF] - 
        █████╗ ███████╗██╗   ██╗██████╗ ███████╗
       ██╔══██╗╚══███╔╝██║   ██║██╔══██╗██╔════╝
       ███████║  ███╔╝ ██║   ██║██████╔╝█████╗
       ██╔══██║ ███╔╝  ██║   ██║██╔══██╗██╔══╝
       ██║  ██║███████╗╚██████╔╝██║  ██║███████╗
       ╚═╝  ╚═╝╚══════╝ ╚═════╝ ╚═╝  ╚═╝╚══════╝

 ██╗ ██████╗ ████████╗    ███████╗██████╗  ██████╗ ███████╗
 ██║██╔═══██╗╚══██╔══╝    ██╔════╝██╔══██╗██╔════╝ ██╔════╝
 ██║██║   ██║   ██║       █████╗  ██║  ██║██║  ███╗█████╗
 ██║██║   ██║   ██║       ██╔══╝  ██║  ██║██║   ██║██╔══╝
 ██║╚██████╔╝   ██║       ███████╗██████╔╝╚██████╔╝███████╗
 ╚═╝ ╚═════╝    ╚═╝       ╚══════╝╚═════╝  ╚═════╝ ╚══════╝

2018-09-22 12:51:34.429 +00:00 [INF] - Installing intermediate certificates.
2018-09-22 12:51:35.093 +00:00 [INF] - Attempting to connect to IoT Hub for client gateway2/$edgeHub via AMQP...
2018-09-22 12:51:35.605 +00:00 [INF] - Connected to IoT Hub for client gateway2/$edgeHub via AMQP, with client operation timeout 60000.
2018-09-22 12:51:35.662 +00:00 [INF] - Created cloud connection for client gateway2/$edgeHub
2018-09-22 12:51:35.672 +00:00 [INF] - New cloud connection created for device gateway2/$edgeHub
2018-09-22 12:51:35.735 +00:00 [INF] - Initializing configuration
2018-09-22 12:51:39.821 +00:00 [INF] - Created persistent store at /tmp/edgeHub
2018-09-22 12:51:40.054 +00:00 [INF] - Created new message store
2018-09-22 12:51:40.068 +00:00 [INF] - Started task to cleanup processed and stale messages
2018-09-22 12:51:40.593 +00:00 [INF] - New device connection for device gateway2/$edgeHub
2018-09-22 12:51:47.335 +00:00 [INF] - Exiting disconnected state
2018-09-22 12:51:47.452 +00:00 [INF] - Entering connected state
2018-09-22 12:51:49.944 +00:00 [INF] - Obtained edge hub config from module twin
2018-09-22 12:51:51.051 +00:00 [INF] - Set the following 2 route(s) in edge hub
2018-09-22 12:51:51.060 +00:00 [INF] - managementToIoTHub: FROM /messages/modules/VibrationModule/outputs/ManagementToCloudOutput INTO $upstream
2018-09-22 12:51:51.061 +00:00 [INF] - uploadToIoTHub: FROM /messages/modules/UploadModule/outputs/ManagementToCloudOutput INTO $upstream
2018-09-22 12:51:51.071 +00:00 [INF] - Updated message store TTL to 14400 seconds
2018-09-22 12:51:51.073 +00:00 [INF] - Updated the edge hub store and forward configuration
2018-09-22 12:51:51.076 +00:00 [INF] - Initialized edge hub configuration
2018-09-22 12:51:51.665 +00:00 [INF] - Starting protocol heads - (MQTT, AMQP, HTTP)
2018-09-22 12:51:51.706 +00:00 [INF] - Starting MQTT head
2018-09-22 12:51:52.517 +00:00 [INF] - Initializing TLS endpoint on port 8883 for MQTT head.
2018-09-22 12:51:52.971 +00:00 [INF] - Starting AMQP head
2018-09-22 12:51:53.044 +00:00 [INF] - Started MQTT head
2018-09-22 12:51:53.744 +00:00 [INF] - Started AMQP head
2018-09-22 12:51:53.763 +00:00 [INF] - Starting HTTP head
2018-09-22 12:51:55.578 +00:00 [INF] - User profile is available. Using '"/home/edgehubuser/.aspnet/DataProtection-Keys"' as key repository; keys will not be encrypted at rest.
2018-09-22 12:51:58.115 +00:00 [WRN] - Overriding address(es) '"http://+:80"'. Binding to endpoints defined in "UseKestrel()" instead.
2018-09-22 12:51:58.376 +00:00 [INF] - Started HTTP head
2018-09-22 12:52:01.999 +00:00 [INF] - Attempting to connect to IoT Hub for client gateway2/VibrationModule via AMQP...
2018-09-22 12:52:02.017 +00:00 [INF] - Connected to IoT Hub for client gateway2/VibrationModule via AMQP, with client operation timeout 60000.
2018-09-22 12:52:02.020 +00:00 [INF] - Created cloud connection for client gateway2/VibrationModule
2018-09-22 12:52:02.021 +00:00 [INF] - New cloud connection created for device gateway2/VibrationModule
2018-09-22 12:52:02.224 +00:00 [INF] - New token requested by client gateway2/VibrationModule, but using existing token as it is usable.
2018-09-22 12:52:02.494 +00:00 [INF] - Successfully generated identity for clientId xlcg001pi03/VibrationModule and username gateway2/gateway2/VibrationModule/?api-version=2018-06-30&DeviceClientType=Microsoft.Azure.Devices.Client%2F1.18.0%20%28.NET%20Core%204.6.0.0%3B%20Linux%204.14.62-v7%2B%20%231134%20SMP%20Tue%20Aug%2014%2017%3A10%3A10%20BST%202018%3B%20Arm%29
2018-09-22 12:52:02.532 +00:00 [INF] - ClientAuthenticated, gateway2/VibrationModule, 4e191a9c
2018-09-22 12:52:02.931 +00:00 [INF] - New device connection for device gateway2/VibrationModule
2018-09-22 12:52:03.007 +00:00 [INF] - Bind device proxy for device gateway2/VibrationModule
2018-09-22 12:52:03.013 +00:00 [INF] - Binding message channel for device Id gateway2/VibrationModule
2018-09-22 12:52:04.102 +00:00 [INF] - Set subscriptions from session state for gateway2/VibrationModule
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc

**Comment by dsajanice** [comment]
Does the client have retries to open a new connection if the connection is dropped?

**Comment by SarithaMaran** [comment]
@carloshm I was also facing the same probelm.
If you set the Environment variable "OptimizeForPerformance" to "false" for the EdgeHub , issue is resolved in my case:
[This is also documented in the Troubleshooting section.](https://docs.microsoft.com/en-us/azure/iot-edge/troubleshoot#stability-issues-on-resource-constrained-devices)

In the portal, this is set in a deployment by selecting "Configure advanced Edge Runtime settings, and adding an environment variable to Edge Hub.

In the deployment JSON, this would look like this:
```

                    "edgeHub": {
                        "type": "docker",
                        "settings": {
                            "image": "mcr.microsoft.com/azureiotedge-hub:1.0",
                            "createOptions": "{\n  \"HostConfig\": {\n    \"PortBindings\": {\n      \"8883/tcp\": [\n        {\n          \"HostPort\": \"8883\"\n        }\n      ],\n      \"443/tcp\": [\n        {\n          \"HostPort\": \"443\"\n        }\n      ],\n      \"5671/tcp\": [\n        {\n          \"HostPort\": \"5671\"\n        }\n      ]\n    }\n  }\n}"
                        },
                        "env": {
                            "OptimizeForPerformance": {
                                "value": "false"
                            }
                        },
                        "status": "running",
                        "restartPolicy": "always"
```

**Comment by carloshm** [comment]
yes, this behavior is with false value. When set to true it is almost impossible to stay connected ;)

**Comment by carloshm** [comment]
> 
> 
> @carloshm I was also facing the same probelm.
> If you set the Environment variable "OptimizeForPerformance" to "false" for the EdgeHub , issue is resolved in my case:
> [This is also documented in the Troubleshooting section.](https://docs.microsoft.com/en-us/azure/iot-edge/troubleshoot#stability-issues-on-resource-constrained-devices)
> 
> In the portal, this is set in a deployment by selecting "Configure advanced Edge Runtime settings, and adding an environment variable to Edge Hub.
> 
> In the deployment JSON, this would look like this:
> 
> ```
> 
>                     "edgeHub": {
>                         "type": "docker",
>                         "settings": {
>                             "image": "mcr.microsoft.com/azureiotedge-hub:1.0",
>                             "createOptions": "{\n  \"HostConfig\": {\n    \"PortBindings\": {\n      \"8883/tcp\": [\n        {\n          \"HostPort\": \"8883\"\n        }\n      ],\n      \"443/tcp\": [\n        {\n          \"HostPort\": \"443\"\n        }\n      ],\n      \"5671/tcp\": [\n        {\n          \"HostPort\": \"5671\"\n        }\n      ]\n    }\n  }\n}"
>                         },
>                         "env": {
>                             "OptimizeForPerformance": {
>                                 "value": "false"
>                             }
>                         },
>                         "status": "running",
>                         "restartPolicy": "always"
> ```

yes, this behavior is with false value. When set to true it is almost impossible to stay connected ;)

**Comment by carloshm** [comment]
> 
> 
> Does the client have retries to open a new connection if the connection is dropped?

We catch expection in the open connection. When there is no connection it throws timeout exception and does not get message.
Don't know which is the edgeHub behavior

**Comment by carloshm** [comment]
@dsajanice  which is the expected behavior? edgeHub stills fails and module could not send any message

**Comment by myagley** [comment]
This looks like the module is failing to connect to the Edge Hub. Is the Edge Hub running and can the module ever connect? I want to rule out any networking issue between the module and the Edge Hub.

What SDK are you using and what is its version (C# 1.18.0, for example)?

**Comment by myagley** [comment]
You may also want to upgrade the Edge Hub (and the rest of the runtime) to 1.0.2. This was just released last Monday and has many improvements for extended offline support. You can read more here: https://azure.microsoft.com/en-us/blog/extended-offline-operation-with-azure-iot-edge/

**Comment by carloshm** [comment]
> 
> 
> This looks like the module is failing to connect to the Edge Hub. Is the Edge Hub running and can the module ever connect? I want to rule out any networking issue between the module and the Edge Hub.
> 
> What SDK are you using and what is its version (C# 1.18.0, for example)?

added in the log lates 1.18 and GA modules for edgeHub and Agent.
We have 7 gateways running this version the module reads information and send it through edgeHub...until at some point, edgeHub fails (as described in this issue) restarting the edgeHub bvrings everything again to life.

@myagley  I would like to understand which is the current behavior for offline capabilities in GA. What we are seeing is when there is any connectivity issue edgeHub fails with timeout, and client connection fails. indeed after those edgeHub restart sometimes it does not handle nicely:

- terminate called after throwing an instance of 'std::bad_alloc'
- what(): std::bad_alloc
or 

- terminate called after throwing an instance of 'std::system_error'
- what(): Resource temporarily unavailable

**Comment by carloshm** [comment]
> 
> 
> You may also want to upgrade the Edge Hub (and the rest of the runtime) to 1.0.2. This was just released last Monday and has many improvements for extended offline support. You can read more here: https://azure.microsoft.com/en-us/blog/extended-offline-operation-with-azure-iot-edge/

@myagley I expected to have an IoT Edge reading signals without connection for several minutes or hours (timeToLiveSec) and when connected every message sent start flowing into IoT Hub. Is this the behavior?

**Comment by baharw** [comment]
I am having the same problem with the exact setup (RPi). Is there a clean solution to this problem?

There are many other folks complaining about the same issue.

**Comment by carloshm** [comment]
We are still experiencing this issue. We manage through exception handling and retrying. Additional log from edgeHub. @myagley  Why edgeHub raise this to the custom module?

2019-01-08 14:23:56.434 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Disposing MessagingServiceClient for device Id gateway001pi01/uploadModule because of exception - Microsoft.Azure.Devices.ProtocolGateway.ProtocolGatewayException: Keep Alive timed out.
2019-01-08 14:23:56.434 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Setting device proxy inactive for device Id gateway001pi01/uploadModule
2019-01-08 14:23:56.435 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Device connection removed for device gateway001pi01/uploadModule
2019-01-08 14:23:56.457 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.EdgeHubConnection] - Updating device gateway001pi01/uploadModule connection status to Disconnected

**Comment by myagley** [comment]
We were just able to repro this behavior with the proper instrumentation in place. This only appears to occur on arm32 devices, and the initial assessment looks like a deadlock in the dotnet core runtime. 

There is a new version of dotnet that came out before the holiday (2.1.6) that may have a fix. We have upgraded to it in #685 and are running another test to see if it still repros.

The symptom when this occurs is that the Edge Hub locks up and refuses any further connections.

**Comment by myagley** [comment]
This fix was released in version `1.0.6` of IoT Edge. Closing this issue for now.
